### PR TITLE
Return 503 for custom resource requests during server start

### LINKED ISSUE #71218: Transient 404 for custom resource after etcd restart
**What happened**:

I got a 404 Not Found error from the apiserver for a custom resource that exists. The 404 error only persisted for one request, after which 200s were returned as normal.

I originally observed this on GKE at the end of a control plane upgrade, but have also reproduced this using https://github.com/kubernetes-sigs/kind and sending SIGTERM to etcd.

**What you expected to happen**:

apiserver shouldn't return 404 for existing resources.

**How to reproduce it (as minimally and precisely as possible)**:

1. Install [kind](https://github.com/kubernetes-sigs/kind#installation-and-usage) and create a cluster.
1. Start a proxy for test requests:
    ```
    KUBECONFIG=$(kind get kubeconfig-path) kubectl proxy &>dev/null &
    ```
1. Start sending test requests to etcd:
    ```
    while true; do
        curl -m 1 -s -D- http://localhost:8005/apis/example.com/v1/namespaces/default/testcrs/my-test-cr
    done | grep ^HTTP | uniq -c
    ```
1. Periodically kill etcd:
    ```
    while true; do
        sudo kill $(ps ax | grep "etcd[ ]--ad" | awk '{print $1}')
        sleep 2m
    done
    ```

Because of the pod restart backoff, it may take some time to reproduce the issue: I ran for about half an hour and saw the following:

```
  16006 HTTP/1.1 200 OK
    618 HTTP/1.1 502 Bad Gateway
      1 HTTP/1.1 404 Not Found
    911 HTTP/1.1 200 OK
   4213 HTTP/1.1 502 Bad Gateway
  17700 HTTP/1.1 200 OK
   6136 HTTP/1.1 502 Bad Gateway
      3 HTTP/1.1 404 Not Found
  50668 HTTP/1.1 200 OK
    997 HTTP/1.1 502 Bad Gateway
      1 HTTP/1.1 404 Not Found
  11817 HTTP/1.1 200 OK
   5023 HTTP/1.1 502 Bad Gateway
      1 HTTP/1.1 404 Not Found
   6367 HTTP/1.1 200 OK
  22570 HTTP/1.1 502 Bad Gateway
```

Note the few `404 Not Found` errors occurring after the `502 Bad Gateway` errors that indicate the unavailability of etcd.

**Anything else we need to know?**:

**Environment**:
- Kubernetes version: v1.12.2 (also observed on GKE with versions 1.10.9 and 1.11.2)
- Cloud provider or hardware configuration: kubeadm amd64
- OS (e.g. from /etc/os-release): Debian Buster-like
- Kernel (e.g. `uname -a`): 4.18.10
- Install tools:
- Others:

/kind bug

### PR DESCRIPTION
**Pull request by liggitt**
**What type of PR is this?**
/kind bug

**What this PR does / why we need it**:

Ensures that requests to potential custom resource endpoints while the server is starting get a 503 error, not a 404

**Which issue(s) this PR fixes**:
Fixes #71218

**Does this PR introduce a user-facing change?**:
```release-note
Resolves a transient 404 response to custom resource requests during server startup
```

### CODE REVIEW

**Comment by liggitt** [comment]
/sig api-machinery
/priority important-soon

will work on an integration test exercising this

**Comment by jpbetz** [comment]
This looks right. +1 for test.

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go [hunk 1]
```diff
@@ -229,6 +231,11 @@ func (r *crdHandler) ServeHTTP(w http.ResponseWriter, req *http.Request) {
 	crdName := requestInfo.Resource + "." + requestInfo.APIGroup
 	crd, err := r.crdLister.Get(crdName)
 	if apierrors.IsNotFound(err) {
+		if !r.hasSynced() {
+			http.Error(w, "server is starting", http.StatusServiceUnavailable)
```
Comment (smarterclayton) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go]: Shouldn’t we return a structured error with a retry after 10s?

**Review by smarterclayton — CHANGES_REQUESTED** [review]

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go (hunk 1 above)
Comment (liggitt) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go]: switched to structured error with timeout

**Comment by liggitt** [comment]
comments addressed, added tests

**Comment by liggitt** [comment]
flake https://github.com/kubernetes/kubernetes/issues/81238

/retest

**Comment by jpbetz** [comment]
Did a pass. The negotiated error handling looks great.

Thanks for the comprehensive testing.

/lgtm

**Pull request merged**